src/nvim/os/shell.c

```diff
@@ -309,15 +318,141 @@ static void system_data_cb(Stream *stream, RBuffer *buf, size_t count,
+    return false;
+  } else if (!visit) {
+    started = os_hrtime();
+  } else if (visit >= max_visits
+             // Gobble up small chunks even if we maxed out. Avoids the case
+             // where the final displayed chunk is very tiny.
+             && (size > 256 || visit >= max_visits + 20)) {
+    uint64_t since = os_hrtime() - started;
+    if (since < NS_1_SECOND) {
+      // Adjust max_visits based on the current relative performance.
+      // Each "pulse" period should last >=1 second so that it is perceptible.
+      max_visits = (2 * max_visits);
+    } else {
+      received = visit = 0;
+      return false;
+    }
+  }
+
+  visit++;
+  // Pulse "..." at the bottom of the screen.
+  size_t tick = (visit == max_visits)
+                ? 3  // Force all dots "..." on last visit.
+                : (visit % 4);
+  pulse_msg[0] = (tick == 0) ? ' ' : '.';
+  pulse_msg[1] = (tick == 0 || 1 == tick) ? ' ' : '.';
+  pulse_msg[2] = (tick == 0 || 1 == tick || 2 == tick) ? ' ' : '.';
+  if (visit == 1) {
+    screen_del_lines(0, 0, 1, (int)Rows, NULL);
+  }
+  int lastrow = (int)Rows - 1;
+  screen_puts_len((char_u *)pulse_msg, ARRAY_SIZE(pulse_msg), lastrow, 0, 0);
+  ui_flush();
+  return true;
+}
+
+/// Saves output in a quasi-ring-buffer. Used to ensure the last ~page of
+/// output for a shell-command is always displayed.
+///
+/// Init mode: Resets the internal state.
+///   output = NULL
+///   size   = 0
+/// Print mode: Displays the current saved data.
+///   output = NULL
+///   size   = SIZE_MAX
+///
+/// @param  output  Data to save, or NULL to invoke a special mode.
+/// @param  size    Length of `output`.
+static void out_data_ring(char *output, size_t size)
+{
+#define MAX_CHUNK_SIZE (OUT_DATA_THRESHOLD / 2)
+  static char    last_skipped[MAX_CHUNK_SIZE];  // Saved output.
+  static size_t  last_skipped_len = 0;
+
+  assert(output != NULL || (size == 0 || size == SIZE_MAX));
+
+  if (output == NULL && size == 0) {          // Init mode
+    last_skipped_len = 0;
+    return;
+  }
+
+  if (output == NULL && size == SIZE_MAX) {   // Print mode
+    out_data_append_to_screen(last_skipped, last_skipped_len, false);
+    return;
+  }
+
+  // This is basically a ring-buffer...
+  if (size >= MAX_CHUNK_SIZE) {               // Save mode
+    size_t start = size - MAX_CHUNK_SIZE;
+    memcpy(last_skipped, output + start, MAX_CHUNK_SIZE);
+    last_skipped_len = MAX_CHUNK_SIZE;
+  } else if (size > 0) {
+    // Length of the old data that can be kept.
+    size_t keep_len   = (MAX_CHUNK_SIZE - size) > last_skipped_len
+                        ? last_skipped_len
+                        : (MAX_CHUNK_SIZE - size);
+    size_t keep_start = last_skipped_len > keep_len
+                        ? last_skipped_len - keep_len
+                        : 0;
+    // Shift the kept part of the old data to the start.
+    memmove(last_skipped, last_skipped + keep_start, keep_len);
+    // Copy the entire new data to the remaining space.
```

Comment:
This could be written a bit shorter:
```c
    size_t keep_len   = MIN(last_skipped_len, MAX_CHUNK_SIZE-size);
    size_t keep_start = last_skipped_len - keep_len;
    
    // Shift the kept part of the old data to the start.
    if (keep_start) {
      memmove(last_skipped, last_skipped + keep_start, keep_len);
    }
```
and making `OUT_DATA_THRESHOLD` unsigned to prevent a compiler warning because of the code change:
```c
#define OUT_DATA_THRESHOLD  1024 * 10U       // 10KB, "a few screenfuls" of data.
```